File: backend/encrypt_env.py
#!/usr/bin/env python3
"""
Encrypt/Decrypt .env files for secure storage
Usage:
    python encrypt_env.py encrypt    # Encrypt .env to .env.enc
    python encrypt_env.py decrypt    # Decrypt .env.enc to .env
"""

import os
import sys
import getpass
import base64
from cryptography.fernet import Fernet
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC


def derive_key(password: str, salt: bytes) -> bytes:
    """Derive encryption key from password using PBKDF2"""
    kdf = PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=32,
        salt=salt,
        iterations=100000,
    )
    key = kdf.derive(password.encode())
    # Fernet requires base64-encoded key
    return base64.urlsafe_b64encode(key)


def encrypt_file(input_file: str, output_file: str, password: str):
    """Encrypt a file using password-based encryption"""
    # Generate random salt
    salt = os.urandom(16)
    
    # Derive key from password
    key = derive_key(password, salt)
    fernet = Fernet(key)
    
    # Read input file
    with open(input_file, 'rb') as f:
        data = f.read()
    
    # Encrypt data
    encrypted_data = fernet.encrypt(data)
    
    # Write salt + encrypted data
    with open(output_file, 'wb') as f:
        f.write(salt + encrypted_data)
    
    print(f"✅ Encrypted {input_file} → {output_file}")


def decrypt_file(input_file: str, output_file: str, password: str):
    """Decrypt a file using password-based encryption"""
    # Read encrypted file
    with open(input_file, 'rb') as f:
        data = f.read()
    
    # Extract salt and encrypted data
    salt = data[:16]
    encrypted_data = data[16:]
    
    # Derive key from password
    key = derive_key(password, salt)
    fernet = Fernet(key)
    
    try:
        # Decrypt data
        decrypted_data = fernet.decrypt(encrypted_data)
        
        # Write decrypted file
        with open(output_file, 'wb') as f:
            f.write(decrypted_data)
        
        print(f"✅ Decrypted {input_file} → {output_file}")
        return True
    
    except Exception as e:
        print(f"❌ Decryption failed: {e}")
        print("Incorrect password or corrupted file")
        return False


def main():
    if len(sys.argv) < 2:
        print("Usage: python encrypt_env.py [encrypt|decrypt]")
        sys.exit(1)
    
    command = sys.argv[1].lower()
    
    if command == "encrypt":
        # Encrypt .env to .env.enc
        if not os.path.exists('.env'):
            print("❌ .env file not found!")
            sys.exit(1)
        
        password = getpass.getpass("Enter encryption password: ")
        confirm = getpass.getpass("Confirm password: ")
        
        if password != confirm:
            print("❌ Passwords don't match!")
            sys.exit(1)
        
        encrypt_file('.env', '.env.enc', password)
        print("\n⚠️  IMPORTANT:")
        print("1. Keep .env.enc in your repository")
        print("2. Add .env to .gitignore (never commit plain .env)")
        print("3. Remember your password - you'll need it to decrypt!")
    
    elif command == "decrypt":
        # Decrypt .env.enc to .env
        if not os.path.exists('.env.enc'):
            print("❌ .env.enc file not found!")
            sys.exit(1)
        
        password = getpass.getpass("Enter decryption password: ")
        
        if decrypt_file('.env.enc', '.env', password):
            print("\n✅ .env file ready to use!")
        else:
            sys.exit(1)
    
    else:
        print(f"❌ Unknown command: {command}")
        print("Usage: python encrypt_env.py [encrypt|decrypt]")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: frontend/encrypt-env.js
#!/usr/bin/env node
/**
 * Encrypt/Decrypt .env files for secure storage (Frontend)
 * Usage:
 *   node encrypt-env.js encrypt    # Encrypt .env.local to .env.local.enc
 *   node encrypt-env.js decrypt    # Decrypt .env.local.enc to .env.local
 */

const crypto = require('crypto');
const fs = require('fs');
const readline = require('readline');

const ALGORITHM = 'aes-256-gcm';
const SALT_LENGTH = 16;
const IV_LENGTH = 16;
const TAG_LENGTH = 16;
const KEY_LENGTH = 32;
const ITERATIONS = 100000;

/**
 * Derive encryption key from password using PBKDF2
 */
function deriveKey(password, salt) {
    return crypto.pbkdf2Sync(password, salt, ITERATIONS, KEY_LENGTH, 'sha256');
}

/**
 * Encrypt file content
 */
function encryptFile(inputFile, outputFile, password) {
    try {
        // Read input file
        const data = fs.readFileSync(inputFile);

        // Generate random salt and IV
        const salt = crypto.randomBytes(SALT_LENGTH);
        const iv = crypto.randomBytes(IV_LENGTH);

        // Derive key from password
        const key = deriveKey(password, salt);

        // Create cipher
        const cipher = crypto.createCipheriv(ALGORITHM, key, iv);

        // Encrypt data
        const encrypted = Buffer.concat([cipher.update(data), cipher.final()]);
        const tag = cipher.getAuthTag();

        // Combine: salt + iv + tag + encrypted data
        const output = Buffer.concat([salt, iv, tag, encrypted]);

        // Write encrypted file
        fs.writeFileSync(outputFile, output);

        console.log(`✅ Encrypted ${inputFile} → ${outputFile}`);
        return true;
    } catch (error) {
        console.error(`❌ Encryption failed: ${error.message}`);
        return false;
    }
}

/**
 * Decrypt file content
 */
function decryptFile(inputFile, outputFile, password) {
    try {
        // Read encrypted file
        const data = fs.readFileSync(inputFile);

        // Extract components
        const salt = data.slice(0, SALT_LENGTH);
        const iv = data.slice(SALT_LENGTH, SALT_LENGTH + IV_LENGTH);
        const tag = data.slice(SALT_LENGTH + IV_LENGTH, SALT_LENGTH + IV_LENGTH + TAG_LENGTH);
        const encrypted = data.slice(SALT_LENGTH + IV_LENGTH + TAG_LENGTH);

        // Derive key from password
        const key = deriveKey(password, salt);

        // Create decipher
        const decipher = crypto.createDecipheriv(ALGORITHM, key, iv);
        decipher.setAuthTag(tag);

        // Decrypt data
        const decrypted = Buffer.concat([decipher.update(encrypted), decipher.final()]);

        // Write decrypted file
        fs.writeFileSync(outputFile, decrypted);

        console.log(`✅ Decrypted ${inputFile} → ${outputFile}`);
        return true;
    } catch (error) {
        console.error(`❌ Decryption failed: ${error.message}`);
        console.error('Incorrect password or corrupted file');
        return false;
    }
}

/**
 * Prompt for password
 */
function promptPassword(prompt) {
    return new Promise((resolve) => {
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });

        // Hide password input
        const stdin = process.stdin;
        stdin.on('data', (char) => {
            char = char.toString();
            if (char === '\n' || char === '\r' || char === '\u0004') {
                stdin.pause();
            } else {
                process.stdout.clearLine();
                process.stdout.cursorTo(0);
                process.stdout.write(prompt + '*'.repeat(rl.line.length));
            }
        });

        rl.question(prompt, (password) => {
            rl.close();
            console.log(''); // New line after password
            resolve(password);
        });
    });
}

/**
 * Main function
 */
async function main() {
    const command = process.argv[2];

    if (!command) {
        console.log('Usage: node encrypt-env.js [encrypt|decrypt]');
        process.exit(1);
    }

    if (command === 'encrypt') {
        // Encrypt .env.local to .env.local.enc
        if (!fs.existsSync('.env.local')) {
            console.error('❌ .env.local file not found!');
            process.exit(1);
        }

        const password = await promptPassword('Enter encryption password: ');
        const confirm = await promptPassword('Confirm password: ');

        if (password !== confirm) {
            console.error('❌ Passwords don\'t match!');
            process.exit(1);
        }

        if (encryptFile('.env.local', '.env.local.enc', password)) {
            console.log('\n⚠️  IMPORTANT:');
            console.log('1. Keep .env.local.enc in your repository');
            console.log('2. Add .env.local to .gitignore (never commit plain .env.local)');
            console.log('3. Remember your password - you\'ll need it to decrypt!');
        } else {
            process.exit(1);
        }
    } else if (command === 'decrypt') {
        // Decrypt .env.local.enc to .env.local
        if (!fs.existsSync('.env.local.enc')) {
            console.error('❌ .env.local.enc file not found!');
            process.exit(1);
        }

        const password = await promptPassword('Enter decryption password: ');

        if (decryptFile('.env.local.enc', '.env.local', password)) {
            console.log('\n✅ .env.local file ready to use!');
        } else {
            process.exit(1);
        }
    } else {
        console.error(`❌ Unknown command: ${command}`);
        console.log('Usage: node encrypt-env.js [encrypt|decrypt]');
        process.exit(1);
    }
}

main().catch(console.error);
